docs/specs/3-13-2023-commands-fit-n-finsh.md

```diff
@@ -0,0 +1,71 @@
+# 3/13/2023 - Commands Fit-and-Finish Proposal
+
+## Design
+
+Commands that interact with the core development workflow will be verb-based. This feels natural to the user as the user is building an application, provisioning infrastructure, and deploying their application to Azure -- which are all actions.
+
+For commands that are of configuration and management in-nature, it is fine to add a noun for a subcommand. For example, `azd env <action>` and `azd config <action>`.
+
+This means that the overall proposed design matches with what `azd` currently offers, and no changes need to occur.
+
+## `azd` usage output
+
+To help introduce user to the command layout, I propose slight changes to the current `azd` usage output (with UX input needed), that looks like the following:
+
+```
+Commands
+  Develop an app
+    init     	: Initialize a new application.
+    provision	: Provision the Azure resources for an application.
+    deploy   	: Deploy the application's code to Azure.
+    up       	: Initialize application, provision Azure resources, and deploy your project with a single command.
+
+  Monitor, test, and release your app
+    monitor  	: Monitor a deployed application.
+    pipeline 	: Manage and configure your deployment pipelines.
+
+  Authenticate with Azure
+    login    	: Log in to Azure.
+    logout   	: Log out of Azure.
+
+  Miscellaneous commands
+    config   	: Manage azd configurations (ex: default Azure subscription, location).
+    down     	: Delete Azure resources for an application.
+    env      	: Manage environments.
+    restore  	: Restore application dependencies.
+    template 	: Find and view template details.
+```
+
+The commands are grouped by their scenarios, with most important scenarios at the top. It is not in alphabetical order, rather the order the user is expected to run the commands. If we added `package` for example, would be after `init`.
+
+## Command changes
+
+The following changes are proposed:
+
+- `azd infra create` and `azd infra delete` should be removed, as the functionality already exists in `azd provision` and `azd down`, and users will find the duplication somewhat confusing.
+- The `--output` flag should be moved to `experimental` (hidden and unsupported unless with feature toggle), with the exception of `azd auth token --output json` which must be supported due to credential handlers.
```

Comment:
I think there are a few other places where we have tooling dependencies in the VS Code extension (`template list` might use the JSON output, and I think there are a few other places) so we need to think about what the implications there are (they can run us with env-vars to turn these on, perhaps?), but I think the general idea of "we don't want to GA the JSON output yet" is solid)

Reply:
Seems like making it just a contract with tooling for now but not a public option is reasonable.

Reply:
Based on that, there are two approaches:

1. We cosmetically hide the flag, but it's available for general usage. No guarantees from the contract, since it is an undocumented flag.
1. We disable the behavior using feature toggle. I think it's unfair for VSCode to now have to reference the feature toggle everytime they use `azd`, and we'll get into trouble with on/off switching (unless we support ephemeral). Even then, it's complicated and unnecessary.

We currently document option 2, but now I'm leaning towards option 1.

Reply:
Option 1 seems right.  If it's hard to hide the flag, I'm also willing for something like "we keep it public but mark it as "preview" and we only do that on the commands we have contracted output for not every command, for now.